File: admin_panel/lib/api/lesson-slides.ts
import { apiFetch, getApiConfig } from "@/lib/api/config";

export interface LessonSlideItem {
  id: string;
  lesson_id: string;
  title: string;
  body: string;
  image_url: string;
  order_no: number;
  is_active: boolean;
  created_at: string;
}

function headers() {
  const { adminApiKey } = getApiConfig();
  return {
    "Content-Type": "application/json",
    ...(adminApiKey ? { "x-admin-api-key": adminApiKey } : {}),
  };
}

async function parseError(response: Response, fallback: string) {
  try {
    const payload = (await response.json()) as { detail?: string };
    return payload.detail?.trim() || fallback;
  } catch {
    return fallback;
  }
}

export async function fetchLessonSlides(lessonId: string) {
  if (!lessonId) return [] as LessonSlideItem[];
  const response = await apiFetch(`/api/v1/lesson-slides?lesson_id=${lessonId}&active_only=false`, {
    cache: "no-store",
    headers: headers(),
  });
  if (!response.ok) throw new Error(await parseError(response, "Slaydlarni olishda xatolik."));
  const data = (await response.json()) as { items: LessonSlideItem[] };
  return data.items ?? [];
}

export async function createLessonSlide(payload: {
  lesson_id: string;
  title: string;
  body: string;
  image_url: string;
  order_no: number;
}) {
  const response = await apiFetch("/api/v1/lesson-slides", {
    method: "POST",
    headers: headers(),
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new Error(await parseError(response, "Slayd qo'shishda xatolik."));
  return (await response.json()) as LessonSlideItem;
}

export async function deleteLessonSlide(id: string) {
  const response = await apiFetch(`/api/v1/lesson-slides/${id}`, {
    method: "DELETE",
    headers: headers(),
  });
  if (!response.ok) throw new Error(await parseError(response, "Slayd o'chirishda xatolik."));
}

File: backend/fastapi/app/schemas/lesson_slides.py
from datetime import datetime

from pydantic import BaseModel, Field


class LessonSlideCreate(BaseModel):
    lesson_id: str = Field(min_length=1)
    title: str = Field(min_length=1, max_length=240)
    body: str = ""
    image_url: str = ""
    order_no: int = Field(default=1, ge=1)


class LessonSlideUpdate(BaseModel):
    title: str | None = None
    body: str | None = None
    image_url: str | None = None
    order_no: int | None = Field(default=None, ge=1)
    is_active: bool | None = None


class LessonSlideItem(BaseModel):
    id: str
    lesson_id: str
    title: str
    body: str
    image_url: str
    order_no: int
    is_active: bool
    created_at: datetime


class LessonSlideListResponse(BaseModel):
    items: list[LessonSlideItem]
